.github/workflows/haskell.yml

```diff
@@ -106,20 +89,110 @@ jobs:
       with:
         path: |
           ${{ steps.setup-haskell.outputs.cabal-store }}
-          dist-newstyle
         # cache is invalidated upon a change to cabal.project (and/or cabal.project.local), a bump to
         # CABAL_CACHE_VERSION or after a week of inactivity
         key: cache-${{ runner.os }}-${{ matrix.ghc }}-${{ env.CABAL_CACHE_VERSION }}-${{ hashFiles('cabal.project*') }}
         # Restoring attempts are from current branch then master
         restore-keys: |
           cache-${{ runner.os }}-${{ matrix.ghc }}-${{ env.CABAL_CACHE_VERSION }}-${{ hashFiles('cabal.project*') }}
 
-    - name: Install dependencies
+    - name: Check workflow test matrix
+      run: cabal build all --dry-run && scripts/check-workflow-test-matrix.hs
+
+    - name: Build dependencies
       run: cabal build all --only-dependencies
 
     - name: Build
       run: cabal build all
 
+    # Pass working directory to test jobs
+    - name: Archive working directory
+      run: tar -cf /var/tmp/state.tar . && mv /var/tmp/state.tar .
+    - name: Upload working directory archive
+      uses: actions/upload-artifact@v4
+      with:
+        name: state-${{ matrix.ghc }}-${{ matrix.os }}
+        path: state.tar
+        overwrite: true
+        retention-days: 1
+
+  test:
+    needs: build
+
```

Comment:
I'm not quite sure how that would be useful. You realize that `complete` will wait for all the ghc versions of `test` to complete, and then run four identical copies of itself?

Reply:
That's exactly what we need. The important part is that it should not run if any of the tests have failed

Reply:
In that case, I don't think it needs a matrix

Reply:
Also, it needs to have some steps or GH will reject it

Reply:
>  I don't think it needs a matrix

Very good point, it'll make it even simpler

Step could be:
```yaml 
run: |
  echo Tada
```

:grin:

Reply:
I've added that (but haven't pushed it yet). However, I'm curious why we couldn't just use the `test` job in the branch protection rules instead of adding another one that doesn't do anything.